### DISCUSSION TITLE
Best practice for mapping values to companion spec

### DISCUSSION THREAD

**Discussion by Fliens** [FreeOpcUa/opcua-asyncio#1304] [Q&A]
Hello I'm using the import_xml() function to load nodes and objects from a opc ua companion spec.
I want to know if you can suggest some best practices to map values from the machine where also the opc ua server is hosted to the nodes generated by the companion spec.

Best regards
Simon

**Comment by schroeder-** [comment]
I don't use any companion spec on a server. But you can study https://github.com/umati/Sample-Server-asyncio which loads multiple companion spec and fill them with data.

**Reply by AndreasHeine** [reply]
the sampleserver has in its origin the idea to record data from an existing machine in csv format and be able to make like a playback from that data to demonstrate a coatingline at Hannover Messe 2021

**Comment by AndreasHeine** [comment]
@Fliens 
whenever you try to map nodes with lets say cache-keys you need to use the ExpandedNodeId-String format e.g "nsu=http://mynamespace.com/UA;i=1001" (the ns-idx is not const enough for a robust long term mapping, it may change over time and would break all you mappings)

the strategies behind it is to resolve the ExpandedNodeId to a regular NodeId after connecting and reading the NamespaceArray (if you are client) or after initialization (if you are server)

you can use a simple python dict to store the values and add callbacks to the nodes
